Extra python dependencies

adriangonz opened the pull request:
This PR partly addresses #564 

## Changes

- `tensorflow` is now an extra package, which can be installed as
  ```console
  $ pip install seldon_core[tensorflow]
  ```
- If `tensorflow` is not present, the `tftensor` data type won't be usable. This is because the `tftensor` type still requires a couple of `tensorflow` methods: `tf.make_ndarray()` and `tf.make_tensor_proto()`.
- Our `seldon_core.proto.prediction_pb2` protobuf still needs the definitions for the `tensorflow` tensor protobuffs: https://github.com/SeldonIO/seldon-core/blob/68c259610a5e93ba1b90547ea9f8d953c997136c/python/seldon_core/proto/prediction_pb2.py#L16-L17
  As a workaround, if `tensorflow` is present, then we don't do anything and we just use the definitions from `tensorflow.core.framework.tensor_pb2`. However, **if `tensorflow` is not present**, we **patch that import path on `seldon_core/proto/__init__.py`** so that it points to our own compiled `tensorflow.core.framework.tensor_pb2` protobuffs.  Patching `syspath` **is not very clean** and can cause problems down the line. I'm still not 100% sure it's a good choice to do this. Suggestions are more than welcome.
- The `Makefile` now compiles the `tensorflow.core.framework.tensor_pb2` protobuffs in case we need to patch the import path.
- `Pillow` has also been moved to the test dependencies, since it's only used on the tests and it's a pretty heavy dependency.

## Notes

As mentioned above, the solution we have implemented is patching `syspath` to point to our own set of pre-compiled `tensorflow.core.framework.tensor_pb2` protobuffs when `tensorflow` is not present. This is required because our own `seldon_core.proto.prediction_pb2` has a hard dependency on them.

A different approach we tried was to keep our set of pre-compiled `tensorflow` protobuffs on a different package which doesn't shadow the `tensorflow.core.framework.tensor_pb2` import path (e.g. a new `tf_proto` package). However, if `tensorflow` is present, it will also try to import its own set of protobuffs and these will then conflict since they have the same ID.

Renaming the package of our pre-compiled `tensorflow.core.framework.tensor_pb2` protobuffs to avoid this conflict doesn't help either, since in that case the `tf.make_ndarray()` and `tf.make_tensor_proto()` methods stop working because the types are different.

Before moving forward, we should consider the pros and cons of overriding `syspath` to make `tensorflow` an optional dependency and perhaps evaluate different alternatives (like writing our own `make_ndarray` and `make_tensor_proto` methods and getting rid of the `tensorflow` dep altogether).

axsaucedo commented:
/uncc @gipster

adriangonz commented:
/assign @axsaucedo @cliveseldon

axsaucedo commented:
/cc @jklaise @cliveseldon

adriangonz commented:
@jklaise @cliveseldon @axsaucedo This solution to make the `tensorflow` dep optional is a bit hacky. Before moving forward, I would love to get your thoughts on this.

axsaucedo commented:
Nice one @adriangonz. As discussed just now it woudl be good to see how complex it woudl be to bring those functions into the codebase, as the Java code already does something similar, where it copies some of the utilities to avoid importing everything. If we can confirm that taking the functions wouldnt' result in a rabbit-hole, then we should go for that. If that is not possible then unless we can think of a different option, this will probably have to be the walkaround.

It woudl also be good to hear @jklaise and @cliveseldon's thoughts on this, in case that point above has already been attempted.

jklaise commented:
I think I agree that copying the utility functions from `tensorflow` would be preferable to patching the path. This would also make the `tftensor` data type usable without tensorflow.

I don't quite understand why the types are different when trying to use tf functions on our pre-compiled protobufs? Is that because the type information includes the package name?

adriangonz commented:
@jklaise that's right. The protobuff types are composed of a package plus an actual type, e.g. `tensorflow.TensorProto` vs `tf_proto.TensorProto`. The classes which `protoc` generates verify that these values follow the correct type.

I started to import the `tensorflow` methods. However, it goes down to C to implement some bits, so it quickly gets complicated. Some examples can be seen on all these [`fast_tensor_util.pyx` routines](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/fast_tensor_util.pyx).

As an alternative, I've started now to rewrite now the "import-shadowing" done on the previous commits to use a (I think) much cleaner option. Instead, I'm going to import all the `tensorflow.core.framework.tensor_pb2` references as `seldon_core.tf_proto.tensor_pb2`. The `seldon_core/tf_proto/__init__.py` file will then decide, based on whether `tensorflow` is present or not, if `seldon_core.tf_proto.tensor_pb2` should lead to `tensorflow.core.framework.tensor_pb2` or our pre-compiled buffers. That way we don't pollute `syspath`.

seldondev commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **adriangonz**

If they are not already assigned, you can assign the PR to them by writing `/assign @adriangonz` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["adriangonz"]} -->

seldondev commented:
@adriangonz: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
serverless-jenkins | b598d13b8c40de539926d567d9074a7c04e7c6e5 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

adriangonz commented:
I've just pushed a change which avoids rewriting `syspath`. Instead, the `__init__.py` file at our pre-compiled `./tensorflow/core/framework/__init__.py` checks if `tensorflow` is present. If it is, it *shadows* our pre-compiled protobuffs with `tensorflow`'s. The tests seem to work both with and without `tensorflow` (skipping the ones using `tf.make_ndarray()` and `tf.make_tensor_proto()` when `tensorflow` is not installed).

axsaucedo commented on python/seldon_core/proto/tensorflow/core/framework/__init__.py:
```diff
@@ -0,0 +1,18 @@
+        If you want to use `tftensor` and Tensorflow's data types
+        install `tensorflow` or install `seldon_core` as
+
+            $ pip install seldon_core[tensorflow]
```
So awesome 👍

axsaucedo commented on python/seldon_core/proto/prediction_pb2.py:
```diff
@@ -731,99 +731,99 @@
-  DESCRIPTOR = _SELDONMESSAGE,
-  __module__ = 'proto.prediction_pb2'
+SeldonMessage = _reflection.GeneratedProtocolMessageType('SeldonMessage', (_message.Message,), {
+  'DESCRIPTOR' : _SELDONMESSAGE,
```
Is this change because we're using an update build of proto or how come?

cliveseldon replied:
yeah, this doesn't seem good if we need to regenerate the proto -> python

adriangonz replied:
@cliveseldon Would it be better to use the build of `protoc` used originally to generate these in case there are breaking changes somewhere?

axsaucedo commented:
Looks solid! Happy with this. I think one way that we could test it is to make sure that the tests in the serverless-jenkins don't run with the side tensorflow, and that it actually installs seldon_core without TF. Perhaps this could be done by removing the TF dependency from https://github.com/adriangonz/seldon-core/blob/extra-python-dependencies/python/requirements.txt

axsaucedo commented:
More specifically, Jenkins X basically runs the jenkins-x.yaml command, which runs make install-dev - this acutally installs it with the requirements.txt, so if you remove the dependency, it should be enough https://github.com/adriangonz/seldon-core/blob/extra-python-dependencies/jenkins-x.yml

axsaucedo commented:
If we can make sure it runs all the tests without issues with a clean environment with the non-TF install, then looks good to me

adriangonz commented:
Good suggestion @axsaucedo! I'll check the changes on the `jenkins-x.yaml`.

seldondev commented:
@adriangonz: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
serverless-jenkins | 9f41504650f4559ff298abd3bd7381ace92b9d0e | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

axsaucedo commented:
/retest

Seems the test itself failed to bind:
```
2019-10-18 16:18:48.746 ERROR 209 --- [         task-1] .a.i.SimpleAsyncUncaughtExceptionHandler : Unexpected exception occurred invoking async method: public void io.seldon.engine.grpc.SeldonGrpcServer.runServer() throws java.lang.InterruptedException,java.io.IOException

java.io.IOException: Failed to bind
        at io.grpc.netty.NettyServer.start(NettyServer.java:246) ~[grpc-netty-1.23.0.jar:1.23.0]
        at io.grpc.internal.ServerImpl.start(ServerImpl.java:180) ~[grpc-core-1.23.0.jar:1.23.0]
        at io.grpc.internal.ServerImpl.start(ServerImpl.java:88) ~[grpc-core-1.23.0.jar:1.23.0]
        at io.seldon.engine.grpc.SeldonGrpcServer.start(SeldonGrpcServer.java:107) ~[classes/:na]
        at io.seldon.engine.grpc.SeldonGrpcServer.runServer(SeldonGrpcServer.java:101) ~[classes/:na]
        at io.seldon.engine.grpc.SeldonGrpcServer$$FastClassBySpringCGLIB$$5bdcc62b.invoke(<generated>) ~[classes/:na]
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) ~[spring-core-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:749) ~[spring-aop-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at org.springframework.aop.interceptor.AsyncExecutionInterceptor.lambda$invoke$0(AsyncExecutionInterceptor.java:115) ~[spring-aop-5.1.9.RELEASE.jar:5.1.9.RELEASE]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_141]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_141]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_141]
        at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_141]
Caused by: java.net.BindException: Address already in use
        at sun.nio.ch.Net.bind0(Native Method) ~[na:1.8.0_141]
        at sun.nio.ch.Net.bind(Net.java:433) ~[na:1.8.0_141]
        at sun.nio.ch.Net.bind(Net.java:425) ~[na:1.8.0_141]
        at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223) ~[na:1.8.0_141]
        at io.netty.channel.socket.nio.NioServerSocketChannel.doBind(NioServerSocketChannel.java:132) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
        at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:551) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1345) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:503) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
        at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:488) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
        at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:984) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
        at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:247) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
        at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:355) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) ~[netty-common-4.1.39.Final.jar:4.1.39.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:416) ~[netty-common-4.1.39.Final.jar:4.1.39.Final]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:515) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918) ~[netty-common-4.1.39.Final.jar:4.1.39.Final]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.39.Final.jar:4.1.39.Final]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.39.Final.jar:4.1.39.Final]
        ... 1 common frames omitted

2019-10-18 16:18:48.924  INFO 209 --- [           main] i.s.engine.config.CustomizationBean      : Customizing EmbeddedServlet
2019-10-18 16:18:48.925  WARN 209 --- [           main] i.s.engine.config.CustomizationBean      : FAILED to find env var [ENGINE_SERVER_PORT], will use defaults for engine server port
2019-10-18 16:18:48.925  INFO 209 --- [           main] i.s.engine.config.CustomizationBean      : setting serverPort[-1]
2019-10-18 16:18:49.879  INFO 209 --- [           main] o.s.s.c.ThreadPoolTaskScheduler          : Initializing ExecutorService 'taskScheduler'
2019-10-18 16:18:49.920  INFO 209 --- [           main] o.s.b.a.e.web.EndpointLinksResolver      : Exposing 3 endpoint(s) beneath base path ''
2019-10-18 16:18:50.007  INFO 209 --- [           main] i.s.e.predictors.RandomABTestUnitTest    : Started RandomABTestUnitTest in 4.011 seconds (JVM running for 43.837)
2019-10-18 16:18:50.017  WARN 209 --- [   scheduling-1] i.s.e.api.rest.SeldonGraphReadyChecker   : Failed to connect to somehost:9000
2019-10-18 16:18:50.018  WARN 209 --- [   scheduling-1] i.s.e.api.rest.SeldonGraphReadyChecker   : Failed to connect to somehost:9000
2019-10-18 16:18:50.018  WARN 209 --- [   scheduling-1] i.s.e.api.rest.SeldonGraphReadyChecker   : Failed to connect to somehost:9000
2019-10-18 16:18:50.018  WARN 209 --- [   scheduling-1] i.s.e.api.rest.SeldonGraphReadyChecker   : Failing somehost:9000
2019-10-18 16:18:50.018  INFO 209 --- [   scheduling-1] i.s.e.api.rest.SeldonGraphReadyChecker   : mean-classifier not ready!
```

seldondev commented:
@adriangonz: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
serverless-jenkins | 9f41504650f4559ff298abd3bd7381ace92b9d0e | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

axsaucedo commented:
Perhaps the changes in the JX file had this effect? You could test by reverting the java piece, but perhaps it's the server

adriangonz commented:
@axsaucedo I think that one doesn't affect the tests. I've seen the binding error on the Java logs before. I haven't digged into it, but it doesn't make the tests fail.

In this case it failed because an error on the python tests:

```python
_____________________ ERROR collecting tests/test_utils.py _____________________
Traceback (most recent call last):
  File "/workspace/source/python/seldon_core/proto/tensorflow/core/framework/__init__.py", line 8, in <module>
    from tensorflow.core.framework import (
ModuleNotFoundError: No module named 'tensorflow'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/workspace/source/python/tests/test_utils.py", line 6, in <module>
    import seldon_core.utils as scu
  File "/workspace/source/python/seldon_core/utils.py", line 11, in <module>
    from seldon_core.proto import prediction_pb2
  File "/workspace/source/python/seldon_core/proto/prediction_pb2.py", line 17, in <module>
    from seldon_core.proto.tensorflow.core.framework import tensor_pb2 as tensorflow_dot_core_dot_framework_dot_tensor__pb2
  File "/workspace/source/python/seldon_core/proto/tensorflow/core/framework/__init__.py", line 18, in <module>
    logger.notice(notice)
AttributeError: 'Logger' object has no attribute 'notice'
```

I was using `logger.notice()` for some unknown reason, which doesn't exist. The thing is that I remember testing that as `logger.info()`. I've changed that, but even then I'm expecting it to fail since there are a couple files which I haven't committed yet which should make the build fail when `tensorflow` is not present.

It's been a really good idea to make Jenkins X test the build path without `tensorflow`!

seldondev commented:
@adriangonz: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
serverless-jenkins | 6e9bc1c15027a4f1ded1c8be2a79fe82c7f3b6fe | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

adriangonz commented:
@axsaucedo @cliveseldon the tests are now added and fixed. I had to change how I was checking if `tensorflow` is present because of [a bug in `v1.14.0` which makes `tensorflow` incompatible with `importlib`](https://github.com/tensorflow/tensorflow/issues/30028).

I've also removed a few no-longer used lines on the `.gitignore` file and added the pre-built TF protobuff interfaces (which were being ignored).

axsaucedo commented:
Looks good from my side. @cliveseldon are we good for landing?

jklaise commented:
/lgtm

seldondev commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/SeldonIO/seldon-core/pull/942#issuecomment-544884576" title="LGTM">jklaise</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **adriangonz**

If they are not already assigned, you can assign the PR to them by writing `/assign @adriangonz` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["adriangonz"]} -->

axsaucedo commented:
/lgtm

seldondev commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/SeldonIO/seldon-core/pull/942#issuecomment-544901972" title="LGTM">axsaucedo</a>*, *<a href="https://github.com/SeldonIO/seldon-core/pull/942#issuecomment-544884576" title="LGTM">jklaise</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [axsaucedo]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

axsaucedo commented:
/hold

cliveseldon commented:
/approve

seldondev commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/SeldonIO/seldon-core/pull/942#issuecomment-544901972" title="LGTM">axsaucedo</a>*, *<a href="https://github.com/SeldonIO/seldon-core/pull/942#issuecomment-544902935" title="Approved">cliveseldon</a>*, *<a href="https://github.com/SeldonIO/seldon-core/pull/942#issuecomment-544884576" title="LGTM">jklaise</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [axsaucedo,cliveseldon]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

axsaucedo commented:
/hold remove
